pkg/operatorstatus/operator_status.go

```diff
@@ -198,13 +200,133 @@ func NewClusterOperatorStatusCondition(conditionType configv1.ClusterStatusCondi
+
+	switch {
+	case availableCondition.Status == "True" && degradedCondition.Status == "False":
+		upgradeableStatus = configv1.ConditionTrue
+	case availableCondition.Status == "False" || degradedCondition.Status == "True":
+		upgradeableStatus = configv1.ConditionFalse
+	}
+
+	upgradeableCondition := newCondition(configv1.OperatorUpgradeable, upgradeableStatus, "", "")
+
+	return []configv1.ClusterOperatorStatusCondition{
+		availableCondition,
+		degradedCondition,
+		progressingCondition,
+		upgradeableCondition,
+	}
+}
+
+// evaluateCondition helps to evaluate conditions and update states.
+func evaluateCondition(co *configv1.ClusterOperator, conditionType configv1.ClusterStatusConditionType, anyMissing *bool, messages *[]string, isDegraded bool) {
+	condition := findCondition(co.Status.Conditions, conditionType)
+	if condition == nil {
+		*anyMissing = true
+		return
+	}
+
+	expectedStatus := configv1.ConditionFalse
+	if isDegraded {
+		expectedStatus = configv1.ConditionTrue
+	}
+
+	if condition.Status == expectedStatus {
+		statusString := fmt.Sprintf("%s=%s", conditionType, condition.Status)
+		*messages = append(*messages, statusString)
+	}
+}
+
+// newAvailableCondition helps to create the Available condition.
+func newAvailableCondition(anyMissing bool, messages []string, releaseVersion string) configv1.ClusterOperatorStatusCondition {
+	if anyMissing {
+		return newCondition(configv1.OperatorAvailable, configv1.ConditionUnknown, "", "")
+	}
+
+	if len(messages) == 0 {
+		return newCondition(configv1.OperatorAvailable, configv1.ConditionTrue, ReasonAsExpected, fmt.Sprintf("Cluster CAPI Operator is available at %s", releaseVersion))
+	}
+
+	return newCondition(configv1.OperatorAvailable, configv1.ConditionFalse, "ControllersNotAvailable", fmt.Sprintf("The following controllers available conditions are not as expected: %s", strings.Join(messages, ", ")))
+}
+
+// newAvailableCondition helps to create the Degraded condition.
+func newDegradedCondition(anyMissing bool, messages []string) configv1.ClusterOperatorStatusCondition {
+	if anyMissing {
+		return newCondition(configv1.OperatorDegraded, configv1.ConditionUnknown, "", "")
+	}
+
+	if len(messages) == 0 {
+		return newCondition(configv1.OperatorDegraded, configv1.ConditionFalse, ReasonAsExpected, "")
+	}
+
+	return newCondition(configv1.OperatorDegraded, configv1.ConditionTrue, ReasonControllersDegraded, fmt.Sprintf("The following controllers degraded conditions are not as expected: %s", strings.Join(messages, ", ")))
+}
+
+// newCondition helps to create a new condition with optional message.
+func newCondition(conditionType configv1.ClusterStatusConditionType, status configv1.ConditionStatus, reason, message string) configv1.ClusterOperatorStatusCondition {
+	return configv1.ClusterOperatorStatusCondition{
+		Type:               conditionType,
+		Status:             status,
+		Reason:             reason,
+		Message:            message,
+		LastTransitionTime: metav1.Now(),
```

Comment:
Does the code applying these take care of making sure `LastTransitionTime` doesn't tick up on every reconcile?